docs/spec/abci/apps.md

```diff
@@ -0,0 +1,207 @@
+# ABCI Applications
+
+Please ensure you've first read the spec for [ABCI Methods and Types](abci.md)
+
+Here we cover the following components of ABCI applications:
+
+- [State](#State) - the interplay between ABCI connections and application state
+  and the differences between `CheckTx` and `DeliverTx`.
+- [Validator Set Updates](#Validator-Set-Updates) - how validator sets are
+  changed during `InitChain` and `EndBlock`
+- [Query](#Query) - standards for using the `Query` method
+- [Crash Recovery](#Crash-Recovery) - handshake protocol to synchronize
+  Tendermint and the application on startup.
+
+## State
+
+Since Tendermint maintains multiple concurrent ABCI connections, it is typical
+for an application to maintain a distinct state for each, and for the states to
+be sycnronized during `Commit`.
+
+### Commit
+
+Before `Commit` is called, Tendermint locks and flushes the mempool so that no new messages will
+be received on the mempool connection. This provides an opportunity to safely update all three
+states to the latest committed state at once.
+
+When `Commit` completes, it unlocks the mempool.
+
+Note that it is not possible to send transactions to Tendermint during `Commit` - if your app
+tries to send a `/broadcast_tx` to Tendermint during Commit, it will deadlock.
```

Comment:
Is the "deadlock" the right word? I thought if somebody's trying to send tx during Commit, it will wait until lock gets released (X ms.) https://github.com/tendermint/tendermint/blob/22445a50297c1c89d48e69b9dd82a080ab7e0209/rpc/core/mempool.go#L52